### PR TITLE
Aggregate cluster view on every node.

### PR DESCRIPTION
**Pull request by zalokhan**
## Overview

Description: Aggregate the complete cluster view on every node.
This is the first step to move towards detecting and addressing partial partitions in the cluster.

Why should this be merged: Every node should have a complete view in order to realize broken links in the cluster.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java [hunk 1]
```diff
@@ -83,15 +87,15 @@
      * In milliseconds.
      */
     @Getter
-    private final long policyExecuteInterval = 1000;
+    private static final long POLICY_EXECUTE_INTERVAL = 1000;
```
[the remark this answers is not in the document]
Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: Probably need to make this configurable.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java [hunk 2]
```diff
@@ -484,7 +490,29 @@ NetworkMetrics getConnectivityView() {
...
+     *
+     * @param pollReport Poll Report to update the current cluster view.
+     */
+    private void updatePeerConnectivityMap(PollReport pollReport) {
+        pollReport.getNodeViewMap().values().stream()
+                .map(nodeView -> nodeView.getNetworkMetrics().getPeerConnectivityView())
+                .flatMap(stringPeerViewMap -> stringPeerViewMap.entrySet().stream())
+                .forEach(stringPeerViewEntry -> clusterView
+                        .compute(stringPeerViewEntry.getKey(), (s, peerView)
+                                -> (peerView == null
```
Comment (StrikingThirteen) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: we might be missing the case when the peerView for a peer received by an intermediary node however the current node (whose management agent is aggregating) view must take precedence. i.e. current node not receiving the heartbeat response from the peer must lead to updating that edge on the graph (cluster view matrix) regardless of intermediary peerView report.

File: runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java [hunk 3]
```diff
@@ -221,16 +223,18 @@ private ClusterStatus getClusterHealth(Layout layout, Set<String> peerResponsive
...
         Set<String> peerResponsiveNodes = new HashSet<>(layout.getAllServers());
-        nodeViewMap.values().stream()
-                .map(NodeView::getNetworkMetrics)
-                // Get all peer connectivity maps from all node views.
-                .map(NetworkMetrics::getPeerConnectivityView)
-                .flatMap(stringBooleanMap -> stringBooleanMap.entrySet().stream())
+        // Truncates all nodes which did not respond to the Heartbeat request.
+        peerResponsiveNodes.retainAll(nodeViewMap.keySet());
+        // Filter out all nodes which are not reachable by the cluster nodes.
+        nodeViewMap.forEach((endpoint, nodeView) -> nodeView.getNetworkMetrics()
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java]: Can we write this in more imperative style. It is hard to read such a long chain of methods.

File: it/src/test/java/org/corfudb/universe/scenario/ScenarioUtils.java [hunk 4]
```diff
@@ -18,6 +19,29 @@
...
 public class ScenarioUtils {
 
+    /**
+     * Refreshes the layout and waits for a limited time for the refreshed layout to
+     * satisfy the expected verifier.
+     *
+     * @param verifier    Layout predicate to test the refreshed layout.
+     * @param corfuClient corfu client.
+     */
+    public static void waitForLayoutChange(Predicate<Layout> verifier, CorfuClient corfuClient) {
```
Comment (Maithem) [inline it/src/test/java/org/corfudb/universe/scenario/ScenarioUtils.java]: There's a similar method that the integration tests use, can you refactor ?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java [hunk 5]
```diff
@@ -157,6 +161,8 @@
     private Set<String> responsiveNodesPeerView = Collections.emptySet();
     // Connectivity of any responsive nodes not responding. Updated by FailureDetector.
     private Set<String> unresponsiveNodesPeerView = Collections.emptySet();
+    // Cluster View compiled by aggregating latest peer views from heartbeat responses.
+    private final ConcurrentHashMap<String, PeerView> clusterView = new ConcurrentHashMap<>();
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: This doesn't seem to be thread safe. Although accessing single entries is thread-safe, making a copy isn't. getConnectivityView can be called from another thread while clusterView is being updated.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java [hunk 6]
```diff
@@ -157,6 +161,8 @@
     private Set<String> responsiveNodesPeerView = Collections.emptySet();
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: Can we encapsulate this information in a class?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java [hunk 7]
```diff
@@ -484,7 +490,29 @@ NetworkMetrics getConnectivityView() {
         // If the management server is not bootstrapped, stamp with INVALID_EPOCH.
         long peerConnectivitySnapshotEpoch = serverContext.getManagementLayout() != null
                 ? serverContext.getManagementLayout().getEpoch() : Layout.INVALID_EPOCH;
-        return new NetworkMetrics(peerConnectivitySnapshotEpoch, peerConnectivityDeltaMap);
+        return new NetworkMetrics(peerConnectivitySnapshotEpoch, ImmutableMap.copyOf(clusterView));
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: Can we come up with a better name than NetworkMetrics?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zalokhan) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: @StrikingThirteen There is only one Management instance per node.
@Maithem I'll make this configurable.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java (hunk 2 above)
Comment (zalokhan) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java]: The latest view of any node view takes precedence regardless of what the current node sees. This is the only way the node can aggregate the cluster view and learn about link failures.

**Comment by zalokhan** [comment]
All comments will be addressed in #1535

**Pull request closed without merging**